discovery/cvm/cvm.go

```diff
@@ -0,0 +1,305 @@
+				labels[cvmLabelTag+labelName] = strPtrToLabelValue(tag.Value)
+				// if key is not valid as label name, encode it to keep all information
+				if !labelName.IsValid() {
+					tagEncoded := strings.ReplaceAll(base32.StdEncoding.EncodeToString([]byte(*tag.Key)), "=", "")
+					labels[cvmLabelTag+model.LabelName(tagEncoded)] = strPtrToLabelValue(tag.Value)
+				}
+			}
+
+			labels[cvmLabelOsName] = strPtrToLabelValue(inst.OsName)
+
+			tg.Targets = append(tg.Targets, labels)
+		}
+
+		if len(resp.Response.InstanceSet) < DefaultPageLimit {
+			break
+		}
+		offset += DefaultPageLimit
+	}
+	return []*targetgroup.Group{tg}, nil
+}
+
+func getClientConfig(config *SDConfig) (*common.Credential, *profile.ClientProfile) {
+	credential := common.NewTokenCredential(
+		defaultWithEnvValues(EnvKeySecretId, config.SecretId),
+		defaultWithEnvValues(EnvKeySecretKey, config.SecretKey),
+		defaultWithEnvValues(EnvKeySecretToken, config.Token),
+	)
+	cpf := profile.NewClientProfile()
+	cpf.HttpProfile.Endpoint = config.Endpoint
+
+	return credential, cpf
+}
+
+func getCvmClient(config *SDConfig) (client *cvm.Client, err error) {
+	credential, cpf := getClientConfig(config)
+	roleArn := defaultWithEnvValues(EnvKeySecretRoleArn, config.RoleArn)
+	if roleArn != "" {
+		// role_arn mode
+		access, err := getUserAccess(config, roleArn)
+		if err != nil {
+			return nil, err
+		}
+		credential = common.NewTokenCredential(*access.TmpSecretId, *access.TmpSecretKey, *access.Token)
+	}
+	return cvm.NewClient(credential, config.Region, cpf)
+}
+
+func getUserAccess(config *SDConfig, roleArn string) (*sts.Credentials, error) {
+	credential, cpf := getClientConfig(config)
+	client, err := sts.NewClient(credential, config.Region, cpf)
+	if err != nil {
+		return nil, err
+	}
+
+	var duration uint64 = 3600
+	request := sts.NewAssumeRoleRequest()
+	request.RoleArn = &roleArn
+	request.DurationSeconds = &duration
+	request.RoleSessionName = common.StringPtr(fmt.Sprintf("%d", time.Now().Unix()))
+	response, err := client.AssumeRole(request)
+	if err != nil {
+		return nil, err
+	}
+	if response == nil || response.Response == nil || response.Response.Credentials == nil {
+		return nil, errors.New("cvm_sd get user access fail")
+	}
+	return response.Response.Credentials, nil
+}
+
+func strPtrSliceToLabelValue(slice []*string, takeOne bool) model.LabelValue {
+	if slice != nil {
+		if takeOne {
+			return model.LabelValue(*slice[0])
+		}
+		return model.LabelValue(strings.Join(common.StringValues(slice), Separator))
+	}
+	return model.LabelValue("")
+}
+
+func strPtrToLabelValue(ptr *string) model.LabelValue {
+	if ptr != nil {
+		return model.LabelValue(*ptr)
+	}
+	return model.LabelValue("")
+}
+
+func intPrtToLabelValue(ptr *int64) model.LabelValue {
+	if ptr != nil {
+		return model.LabelValue(fmt.Sprintf("%d", *ptr))
+	}
+	return model.LabelValue("")
+}
+
+// if defaultValue is blank, use env values
+func defaultWithEnvValues(key, defaultValue string) string {
```

Comment:
Prometheus does not allow to use environment variables for configuration.

Reply:
fixed